### PR TITLE
Non-record: Add submission track_non_record_16mb/2026-03-23_DepthRecurrent_TTT

### PR DESCRIPTION
**Pull request by SergiuDeveloper**
# Depth-Recurrent Test-Time Training

This record captures a non-record submission having the following particularities: depth recurrence, and chunk-causal test-time LoRA adaptation at evaluation.

## Motivation

Under the 16MB artifact cap, the model size is tightly constrained. There is little room to add layers or width. This submission explores a complementary axis: instead of making the model bigger through depth, make it bigger through depth recurrence. Depth recurrence runs each encoder and decoder block multiple times before propagating activations forward, effectively increasing representational depth without adding parameters. At evaluation time, chunk-causal TTT LoRA adapters let the model specialize to each document's distribution before predicting it, refining one chunk at a time.

## Configuration

- **Track:** `non-record-16mb`, 600s wallclock cap, under the `16,000,000` byte artifact cap
- **Layout:** `VOCAB_SIZE=1024 NUM_LAYERS=9 MODEL_DIM=512 NUM_HEADS=8 NUM_KV_HEADS=4 MLP_MULT=2`
- **Tied embeddings:** `TIE_EMBEDDINGS=1`
- **Depth recurrence:** `ENCODER_LOOPS=2 DECODER_LOOPS=2`
- **Batching:** `TRAIN_BATCH_TOKENS=1572864 TRAIN_SEQ_LEN=1024`
- **TTT LoRA:** `TTT_LORA_RANK=8 TTT_CHUNK_SIZE=256`

## Running

```bash
MAX_WALLCLOCK_SECONDS=600 \
VAL_LOSS_EVERY=0 \
torchrun --standalone --nproc_per_node=8 records/track_non_record_16mb/2026-03-23_DepthRecurrent_TTT/train_gpt.py
```

## Key Metrics

Reported metrics are **averages across three seeds** (1337, 42, 404).

- Post-quant TTT LoRA eval: `val_loss: 2.0416`, `val_bpb: 1.2092`
- Pre-quant eval: `val_loss: 2.0950`, `val_bpb: 1.2408`
- Total submission size: `15,544,590 bytes`
- Training stopped at `2570` steps due to the 600s wallclock cap

## Included Files

- `train_gpt.py` (code snapshot used for the run)
- `submission.json` (leaderboard metadata)
- 3 run log files for each seed

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: Add submission track_non_record_16mb/2026-03-23_DepthRecurrent_TTT

**Compliance:** LOOKS CLEAN — legal score-first-per-chunk TTT (PR #1413 pattern)


## PR #495 — Non-record: track_non_record_16mb/2026-03-23_DepthRecurrent_TTT

**Head SHA:** 1e72866037bcad06c6c0ea145a736d16359260d5  
**File reviewed:** `records/track_non_record_16mb/2026-03-23_DepthRecurrent_TTT/train_gpt.py` (1383 lines)

---

### Check 1: N-gram family bug (CLOSE trigger)
No n-gram, hash-table, bigram, or trigram code found anywhere in the file. The target token never appears in a hash lookup key. **PASS — no violation.**

### Check 2: Pre-Quant TTT (CLOSE trigger)
The TTT evaluation function `eval_val_ttt_lora` uses `torch.optim.Adam` (not AdamW) with a single gradient step per chunk (line 956: `cur_opt.step()`). There is no multi-epoch loop over val_tokens. The function is called at line 1369, after the model has already been quantized and dequantized (lines 1320–1343), so the scored TTT run is post-quant, not pre-quant. **PASS — no violation.**

### Check 3: Legal TTT (CLEAN pattern)
The TTT loop in `eval_val_ttt_lora` follows strict score-first-per-chunk order:
- Lines 940–948: `_accumulate_bpb` scores the current chunk (no-grad) **before** any parameter update.
- Lines 950–956: Adam step trains LoRA on the current chunk's loss only after scoring is complete.
- Final chunk (`ci == max_nc - 1`) never triggers a train step (`needs_train` is False when all docs are on their last chunk).

This matches the legal TTT pattern from PR #1413. **CLEAN.**

### Check 4: Scored-region SLOT
No scored-region manipulation, custom BPB weighting, or region-gating logic found. The `_accumulate_bpb` helper is a straightforward per-token loss accumulation without any selective exclusion of tokens or regions. **No SLOT flag warranted.**

### Check 5: Pure neural
Architecture is a standard transformer (9 layers, dim 512) with depth recurrence (`ENCODER_LOOPS`, `DECODER_LOOPS` loop each block 2×) plus LoRA adapters at eval time. No lookup tables, n-gram caches, or non-neural scoring components. **CLEAN.**

---

**Verdict:** LOOKS CLEAN — legal TTT implementation.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source. If this review misread your code, please call it out so I can re-audit manually.*
